src/runtime/heap/mcache.c

```diff
@@ -41,19 +42,29 @@ u64 mcache_alloc(heap h, bytes b)
     rputs(": ");
 #endif
     if (b > m->parent_threshold) {
+        if (!m->fallbacks) {
+            m->fallbacks = allocate_table(m->meta, identity_key, pointer_equal);
```

Comment:
Thanks for pointing this out; I completely overlooked it. I tried a few different approaches to formulating a key and ran some mock allocation tests in the kernel. Even shifting down by PAGELOG wouldn't be an improvement for our primary use, because it would still produce the same key whenever the allocation size is over 16KB (4*PAGESIZE), and the minimum fallback size for (non-lowmem) Nanos general mcache is 128KB. As allocation sizes increase, the address lsb goes even higher, so it seemed the right thing would be to mix every pair (default four buckets) of bits starting at bit 12 to some point. I chose 23 as the high bit as this helps capture some entropy from 2MB size allocations as well. In a test of allocating memory from 128KB through 4MB, this produced an even distribution among the four buckets.

Incidentally, it's kind of too bad the table interface doesn't use closures for the key and compare, because then then the key function could be tuned to the specific mcache. But I think the above, while adding 15 instructions on x86 (many of which I expect would issue in parallel), is sufficiently general.